File: ros2_ws/src/lidar_perception/lidar_perception/lidar_perception.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy
from sensor_msgs.msg import PointCloud2
from sensor_msgs_py import point_cloud2
import numpy as np
import time
import ros2_numpy as rnp

class LidarPerception(Node):
    def __init__(self):
        super().__init__('lidar_perception')
        self.latest_msg = None
        self.increment = 0
        self.total_points = None
        qos = QoSProfile(depth = 10)
        qos.reliability = ReliabilityPolicy.BEST_EFFORT
        self.pub = self.create_publisher(PointCloud2,'/processed_points',qos)
        self.sub = self.create_subscription(PointCloud2,'/pandar_points',self.callback,qos)
        self.timer = self.create_timer(0.5,self.timer_callback)



    def callback(self, msg):
        self.latest_msg = msg
        self.increment += 1
        self.total_points = msg.height*msg.width
        self.get_logger().info(f"No. of points are {self.total_points}")
        

    def timer_callback(self):
        if self.latest_msg == None:
            self.get_logger().info("Waiting for pcl msg")
            return
        msg = self.latest_msg
        t1 = time.time()
        points = ros_numpy.point_cloud2.pointcloud2_to_array(msg)
        t2 = time.time()
        self.get_logger().info(f"shape of points array is {points.shape}")
        Distance = np.sqrt(points[:,0]**2 + points[:,1]**2 + points[:,2]**2)
        points = points[Distance < 20]
        points = list(map(list,points))

        
        t3 = time.time()
        self.processed_pcl = point_cloud2.create_cloud(msg.header,msg.fields,points)
        t4 = time.time()
        self.pub.publish(self.processed_pcl)
        t5 = time.time()
        self.get_logger().info(f"""time to read data from pcl is {t2 -t1} 
        time to filter pcl data is {t3-t2} 
        time to make new pcl msg from filtered data is {t4-t3} 
        time to publish is {t5-t4}""")
    
def main(args = None):
    rclpy.init(args = args)
    node = LidarPerception()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: ros2_ws/src/lidar_perception_cpp/src/lidar_perception.cpp
#include "lidar_perception_cpp/lidar_perception.hpp"

LidarPerceptionCpp::LidarPerceptionCpp():Node("lidar_perception_node")
{   
    rclcpp::QoS qos(rclcpp::KeepLast(10));

    // Optionally set reliability
    qos.reliability(RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT);

    pub = this->create_publisher<sensor_msgs::msg::PointCloud2>("/processed_points",qos);
    sub = this->create_subscription<sensor_msgs::msg::PointCloud2>("/pandar_points",qos,std::bind(&LidarPerceptionCpp::topic_callback,this,std::placeholders::_1));

}


void LidarPerceptionCpp::topic_callback(const sensor_msgs::msg::PointCloud2::SharedPtr msg){
    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud(new pcl::PointCloud<pcl::PointXYZ>);
    pcl::fromROSMsg(*msg, *cloud);
    RCLCPP_INFO(this->get_logger(), "Original cloud size: %zu", cloud->points.size());

    pcl::PointCloud<pcl::PointXYZ>::Ptr cloud_filtered(new pcl::PointCloud<pcl::PointXYZ>);

    for(const auto &point : cloud->points){
        float distance = std::sqrt(point.x*point.x + point.y*point.y + point.z*point.z);
        if (distance <= 20){
            cloud_filtered->points.push_back(point);
        }
        
    }
    cloud_filtered->width = cloud_filtered->points.size();
        cloud_filtered->height = 1;  // unorganized cloud
        cloud_filtered->is_dense = true;
        // Create ROS2 message
        sensor_msgs::msg::PointCloud2 output_msg;

        // Convert PCL cloud to ROS2 PointCloud2
        pcl::toROSMsg(*cloud_filtered, output_msg);

        // Keep the same header (frame_id and timestamp)
        output_msg.header = msg->header;
        pub->publish(output_msg);

}

int main(int argc, char * argv[])
{
    rclcpp::init(argc,argv);
    rclcpp::spin(std::make_shared<LidarPerceptionCpp>());
    rclcpp::shutdown();
    return 0;
}